plugins/bundle/plugin.go

```diff
@@ -0,0 +1,282 @@
+// Copyright 2018 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+// Package bundle implements bundle downloading.
+package bundle
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/bundle"
+	"github.com/open-policy-agent/opa/plugins"
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/util"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// min amount of time to wait following a failure
+	minRetryDelay = time.Millisecond * 100
+)
+
+var (
+	defaultMinDelay = int64(time.Second * 60)
+	defaultMaxDelay = int64(time.Second * 120)
+)
+
+// PollingConfig represents configuration for the plugin's polling behaviour.
+type PollingConfig struct {
+	MinDelaySeconds *int64 `json:"min_delay_seconds,omitempty"` // min amount of time to wait between successful poll attempts
+	MaxDelaySeconds *int64 `json:"max_delay_seconds,omitempty"` // max amount of time to wait between poll attempts
+}
+
+// Config represents configuration the plguin.
+type Config struct {
+	Name    string        `json:"name"`
+	Service string        `json:"service"`
+	Polling PollingConfig `json:"polling"`
+}
+
+func (c *Config) validateAndInjectDefaults(services []string) error {
+
+	if c.Name == "" {
+		return fmt.Errorf("invalid bundle name %q", c.Name)
+	}
+
+	found := false
+
+	for _, svc := range services {
+		if svc == c.Service {
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		return fmt.Errorf("invalid service name %q in bundle %q", c.Service, c.Name)
+	}
+
+	if c.Polling.MinDelaySeconds == nil && c.Polling.MaxDelaySeconds == nil {
+		c.Polling.MinDelaySeconds = &defaultMinDelay
+		c.Polling.MaxDelaySeconds = &defaultMaxDelay
+	} else if c.Polling.MinDelaySeconds != nil && c.Polling.MaxDelaySeconds == nil {
+		c.Polling.MaxDelaySeconds = c.Polling.MinDelaySeconds
+	} else if c.Polling.MinDelaySeconds == nil && c.Polling.MaxDelaySeconds != nil {
+		c.Polling.MinDelaySeconds = c.Polling.MaxDelaySeconds
+	} else if *c.Polling.MaxDelaySeconds < *c.Polling.MinDelaySeconds {
+		return fmt.Errorf("max polling delay must be <= min polling delay in bundle %q", c.Name)
```

Comment:
max polling delay must be "**>=**" min polling delay in bundle

Reply:
Fixed.